lib/pbench/server/api/resources/query_controllers.py

```diff
@@ -0,0 +1,273 @@
+from flask import request
+from flask_restful import Resource, abort
+import requests
+
+from dateutil import parser, rrule
+
+
+class QueryControllers(Resource):
+    """
+    Abstracted Pbench API to get date-bounded controller data.
+    """
+
+    def __init__(self, app, api):
+        self.app = app
+        self.api = api
+
+    @staticmethod
+    def _genMonthUriRange(prefix, index, start, end):
+        """
+        _genMonthUriRange Construct a comma-separated list of index names
+        qualified by year and month suitable for use in the Elasticsearch
+        /_search query URI.
+
+        The month is incremented by 1 from "start" to "end"; for example,
+
+        _genMonthUriRange('drb.', 'v4.run.', '2020-08', '2020-10') will result
+        in
+
+            'drb.v4.run.2020-08,drb.v4.run.2020-09,drb.v4.run.2020-10,'
+
+        This code is adapted from Javascript in the Pbench dashboard's
+        'moment_constants.js' getAllMonthsWithinRange function.
+
+        Args:
+            prefix ([string]): The Pbench server index prefix
+            index ([string]): The desired monthly index root
+            start ([datetime.datetime]): The start time
+            end ([datetime.datetime]): The end time
+
+        Returns:
+            [string]: A comma-separated list of month-qualified index names
+        """
+        monthResults = list()
+        queryString = ""
+        for m in rrule.rrule(rrule.MONTHLY, dtstart=start, until=end):
```

Comment:
Does `rrule` validate that `start` and `end` are valid dates to begin with?

Reply:
`__getMonthUriRange` takes `datetime.datetime` objects, not raw strings; validation is done in the caller:

```
try:
    start = parser.parse(start_arg)
    end = parser.parse(end_arg)
except Exception as e:
    [...] abort(400, ...)
```